File: nanocli/commands/review.py
import typer
from nanocli.llm import LLMService
from nanocli.files import FileHandler
from nanocli.memory import MemoryManager
from rich.console import Console

console = Console()

def review(
    file_path: str = typer.Argument(..., help="Path ke file yang ingin di-review"),
    model: str = typer.Option(None, "--model", "-m", help="Model yang digunakan"),
    deep: bool = typer.Option(False, "--deep", help="Gunakan model deep review"),
    project: bool = typer.Option(False, "--project", help="Gunakan Project Memory"),
):
    """
    Review kode pada file tertentu.
    """
    llm = LLMService()
    file_handler = FileHandler()
    
    content = file_handler.read_file(file_path)
    if not content:
        return

    selected_model = model
    num_ctx = None
    
    if not selected_model:
        if deep:
            selected_model = llm.config.models.deep
            num_ctx = llm.config.ollama.num_ctx_deep
        else:
            selected_model = llm.config.models.default
            num_ctx = llm.config.ollama.num_ctx_normal

    system_prompt = (
        "You are a strict but practical code reviewer. "
        "Review the file based on correctness, maintainability, security, performance, and consistency. "
        "Return concise findings with location, issue, impact, and suggested fix. "
        "Do not rewrite the whole file unless requested."
    )
    
    if project:
        memory = MemoryManager()
        context = memory.get_project_context()
        relevant_mem = memory.get_relevant_memory()
        if context or relevant_mem:
            system_prompt += "\n\nProject Context for consistency:\n"
            if context: system_prompt += context
            if relevant_mem: system_prompt += relevant_mem
            console.print("[dim]Project Memory aktif untuk review.[/dim]")

    user_prompt = f"Review file berikut:\n\nPath: {file_path}\n\n```\n{content}\n```"
    
    messages = [
        {"role": "system", "content": system_prompt},
        {"role": "user", "content": user_prompt}
    ]
    
    console.print(f"[dim]Reviewing {file_path} using {selected_model}...[/dim]")
    
    response_gen = llm.chat(messages, model=selected_model, num_ctx=num_ctx)
    if response_gen:
        llm.stream_response(response_gen)

File: NanoCLI/src/commands/review.ts
import chalk from 'chalk';
import { ConfigManager } from '../files/configManager';
import { OpenRouterClient, Message } from '../llm/openrouterClient';
import { MemoryManager } from '../memory/memoryManager';
import { TokenBudgetManager } from '../tokens/tokenBudgetManager';
import { ContextCompactor } from '../context/contextCompactor';
import { StatsManager } from '../tokens/statsManager';
import { ModelManager } from '../llm/modelManager';
import { Renderer } from '../ui/render';
import { isValidMode, VALID_MODES } from '../config/modes';
import { safeReadTextFile } from '../files/safeFileReader';

export class ReviewCommand {
  private configManager: ConfigManager;
  private memoryManager: MemoryManager;
  private tokenManager: TokenBudgetManager;
  private compactor: ContextCompactor;
  private statsManager: StatsManager;
  private modelManager: ModelManager;

  constructor(projectRoot: string = process.cwd()) {
    this.configManager = new ConfigManager(projectRoot);
    this.memoryManager = new MemoryManager(projectRoot);
    this.tokenManager = new TokenBudgetManager();
    this.compactor = new ContextCompactor();
    this.statsManager = new StatsManager(projectRoot);
    this.modelManager = new ModelManager(projectRoot);
  }

  async execute(filePath: string, options: any) {
    const mode = options.mode || 'normal';

    if (!isValidMode(mode)) {
      Renderer.printStatus(`Mode tidak valid: "${mode}". Pilih salah satu: ${VALID_MODES.join(', ')}`, 'error');
      return;
    }

    const apiKey = await this.configManager.getApiKey();
    const modelId = options.model || await this.configManager.getModelForMode(mode);

    if (!apiKey) {
      Renderer.printStatus('API Key tidak ditemukan. Jalankan "nanocli setup" terlebih dahulu.', 'error');
      return;
    }

    let fileResult;
    try {
      fileResult = await safeReadTextFile(filePath, {
        projectRoot: process.cwd(),
        maxBytes: 100_000,
      });
    } catch (err: any) {
      Renderer.printStatus(err.message, 'error');
      return;
    }

    const { content: fileContent, relativePath: fileName } = fileResult;

    const client = new OpenRouterClient(apiKey);
    const messages: Message[] = [];

    // 1. System Prompt for Review
    messages.push({
      role: 'system',
      content: `You are an expert code reviewer. Review the provided code for:
- Correctness and potential bugs
- Security vulnerabilities
- Performance optimizations
- Maintainability and best practices
- Consistency with common coding standards

Provide your feedback in a structured Markdown format with clear headings.`
    });

    // 2. Project Context — enriched query: coding style, patterns, known issues
    const fileExt = fileName.split('.').pop() ?? '';
    const langMap: Record<string, string> = {
      ts: 'TypeScript', tsx: 'TypeScript React', js: 'JavaScript',
      py: 'Python', go: 'Go', rs: 'Rust', php: 'PHP',
    };
    const lang = langMap[fileExt] ?? fileExt;
    const enrichedReviewQuery = [
      `code review style patterns best practices ${lang}`,
      `architecture decision convention rule`,
      `file ${fileName}`,
    ].join(' ');
    const context = await this.memoryManager.getContextForQuery(enrichedReviewQuery);
    if (context) {
      messages.push({
        role: 'system',
        content: `Project Context (coding style, decisions, known bugs):\n${context}`
      });
    }

    // 3. Add File Context
    messages.push({
      role: 'user',
      content: `Please review this file: ${fileName}\n\nContent:\n\`\`\`\n${fileContent}\n\`\`\``
    });

    // 4. Fetch model metadata dan compact dengan model-aware budget
    const modelMetadata = await this.modelManager.getModel(modelId);
    const compactedMessages = this.compactor.compactMessages(messages, mode, modelMetadata?.context_length);

    // 5. Cost Guard Info
    if (modelMetadata) {
      const inputTokens = this.tokenManager.countMessageTokens(compactedMessages);
      const estimatedCost = this.tokenManager.estimateCost(inputTokens, modelMetadata.pricing, 1000);
      if (estimatedCost > 0.05) {
        Renderer.printStatus(`Estimasi biaya review: $${estimatedCost.toFixed(4)} USD.`, 'warn');
      }
    }

    Renderer.printStatus(`Sedang me-review ${chalk.bold(fileName)}...`, 'info');

    // 5. Stream Response
    process.stdout.write(chalk.cyan('\nReview Result:\n'));
    let fullResponse = '';
    try {
      const stream = client.streamChat({
        model: modelId,
        messages: compactedMessages,
        stream: true
      });

      for await (const chunk of stream) {
        process.stdout.write(chunk);
        fullResponse += chunk;
      }
      process.stdout.write('\n\n');

      // 6. Log Usage
      if (modelMetadata) {
        const inputTokens = this.tokenManager.countMessageTokens(compactedMessages);
        const outputTokens = this.tokenManager.countTextTokens(fullResponse);
        const costUsd = this.tokenManager.estimateCost(inputTokens, modelMetadata.pricing, outputTokens);

        await this.statsManager.logUsage({
          timestamp: Date.now(),
          modelId,
          mode,
          inputTokens,
          outputTokens,
          costUsd
        });
      }
    } catch (error: any) {
      Renderer.printStatus(error.message, 'error');
    }
  }
}
